### PR TITLE
fix warnings in gcc-8.1+

### LINKED ISSUE #4196: bootloader fails to build on Alpine
* What is the exact problem?

I am attempting to build PyInstaller 3.4 from source in an Alpine Linux container. The  build fails with the following errors: 
```
Waf: Entering directory `/tmp/pyinstaller/bootloader/build/debug'
[ 1/13] Compiling src/pyi_python.c
[ 2/13] Compiling src/pyi_global.c
[ 3/13] Compiling src/pyi_archive.c
[ 4/13] Compiling src/main.c
[ 5/13] Compiling src/pyi_utils.c
[ 6/13] Compiling src/pyi_path.c
[ 7/13] Compiling src/pyi_win32_utils.c
[ 8/13] Compiling src/pyi_pythonlib.c
[ 9/13] Compiling src/pyi_main.c
../../src/pyi_path.c: In function 'pyi_path_join':
../../src/pyi_path.c:143:5: error: 'strncpy' specified bound depends on the length of the source argument [-Werror=stringop-overflow=]
     strncpy(result, path1, strlen(path1));
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/pyi_path.c:156:9: error: 'strncat' specified bound depends on the length of the source argument [-Werror=stringop-overflow=]
         strncat(result, path2, len - 2);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/pyi_path.c:152:11: note: length computed here
     len = strlen(path2);
           ^~~~~~~~~~~~~
cc1: all warnings being treated as errors

../../src/pyi_pythonlib.c: In function 'pyi_pylib_start_python':
../../src/pyi_pythonlib.c:469:9: error: 'strncat' specified bound 1 equals source length [-Werror=stringop-overflow=]
         strncat(pypath, PYI_SEPSTR, strlen(PYI_SEPSTR));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/pyi_pythonlib.c:470:9: error: 'strncat' specified bound 16 equals source length [-Werror=stringop-overflow=]
         strncat(pypath, "base_library.zip", strlen("base_library.zip"));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/pyi_pythonlib.c:471:9: error: 'strncat' specified bound 1 equals source length [-Werror=stringop-overflow=]
         strncat(pypath, PYI_PATHSEPSTR, strlen(PYI_PATHSEPSTR));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/pyi_pythonlib.c:468:9: error: 'strncpy' specified bound depends on the length of the source argument [-Werror=stringop-overflow=]
         strncpy(pypath, status->mainpath, strlen(status->mainpath));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/pyi_pythonlib.c:472:9: error: 'strncat' specified bound depends on the length of the source argument [-Werror=stringop-overflow=]
         strncat(pypath, status->mainpath, strlen(status->mainpath));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/pyi_pythonlib.c:464:9: error: 'strncpy' specified bound depends on the length of the source argument [-Werror=stringop-overflow=]
         strncpy(pypath, status->mainpath, strlen(status->mainpath));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

Waf: Leaving directory `/tmp/pyinstaller/bootloader/build/debug'
Broken revdeps detected on {
	{task 140266198207176: strip run_d -> },
	{task 140266198207064: cprogram main.c.1.o,pyi_archive.c.1.o,pyi_global.c.1.o,pyi_launch.c.1.o,pyi_main.c.1.o,pyi_path.c.1.o,pyi_python.c.1.o,pyi_python27_compat.c.1.o,pyi_pythonlib.c.1.o,pyi_utils.c.1.o,pyi_win32_utils.c.1.o -> run_d}}
The command '/bin/sh -c git clone --depth 1 --single-branch --branch v3.4 https://github.com/pyinstaller/pyinstaller.git /tmp/pyinstaller     && cd /tmp/pyinstaller/bootloader     && python ./waf configure --no-lsb all     && pip install ..     && rm -Rf /tmp/pyinstaller' returned a non-zero code: 1
```

* How to reproduce

1. Clone this repo: https://github.com/six8/pyinstaller-alpine
2. Run build.sh

It doesn't seem to matter what base image I use for alpine for this error.

* Further down the spiral

Originally I ran into this issue: https://github.com/gliderlabs/docker-alpine/issues/48

What would it take to get out of the box support for Alpine and Python 3.7 from pyinstaller so that this workaround isn't needed, and one can simply `pip install pyimage` in an alpine DockerFile?

### PR DESCRIPTION
**Pull request by linxon**
Fixes #4196

### CODE REVIEW

File: bootloader/src/pyi_path.c [hunk 1]
```diff
@@ -153,6 +154,7 @@ pyi_path_join(char *result, const char *path1, const char *path2)
 
     if (path2[len - 1] == PYI_SEP) {
         /* Append path2 without slash. */
+        *(result + len) = '\0';
```
Comment (htgoebel) [inline bootloader/src/pyi_path.c]: How does this avoid the warning?
For my understanding, this line should go behind the next one to *really* ensure the string is \o-terminated.

File: bootloader/src/pyi_pythonlib.c [hunk 2]
```diff
@@ -461,15 +461,15 @@ pyi_pylib_start_python(ARCHIVE_STATUS *status)
...
-        strncpy(pypath, status->mainpath, strlen(status->mainpath));
-        strncat(pypath, PYI_SEPSTR, strlen(PYI_SEPSTR));
-        strncat(pypath, "base_library.zip", strlen("base_library.zip"));
-        strncat(pypath, PYI_PATHSEPSTR, strlen(PYI_PATHSEPSTR));
-        strncat(pypath, status->mainpath, strlen(status->mainpath));
+        memcpy(pypath, status->mainpath, strlen(status->mainpath));
+        memcpy(pypath, PYI_SEPSTR, strlen(PYI_SEPSTR));
+        memcpy(pypath, "base_library.zip", strlen("base_library.zip"));
+        memcpy(pypath, PYI_PATHSEPSTR, strlen(PYI_PATHSEPSTR));
+        memcpy(pypath, status->mainpath, strlen(status->mainpath));
```
Comment (htgoebel) [inline bootloader/src/pyi_pythonlib.c]: I doubt, this code is correct, as successive calls to `memcpy` overwrite the previously copied bytes. I suggest code along this lines :
```
        size_t len = pypath_may - strlen(string);
        strncpy(pypath, status->mainpath, len);
        len -= strlen(PYI_SEPSTR);
        strncat(pypath, PYI_SEPSTR, len);
...
```

**Comment by htgoebel** [comment]
@linxon Are you going to apply the requested changes soon? Otherwise I'll cose the pll-request as being stale.

**Comment by linxon** [comment]
@htgoebel unfortunately, I no have time in the current moment.

P.S. 
it was my hopeless attempt to fix this error. I'm not sure I can handle it. 
I can't try to use your changes but I can't test this on my machine in the current moment

Sorry

**Comment by htgoebel** [comment]
@linxon Sad to hear/read. I'll cloe the pull-requet them, since it does not make sense to keep pull-requests open wich will not get finished soon. Hopefully someoen else will pick up your work. Thansk

**Pull request closed without merging**